File: photonflow-cpp/src/blocks/measurement/osa_probe.cpp
/**
 * @file osa_probe.cpp
 * @brief OSAProbe implementation.
 */

#include "photonflow/blocks/measurement/osa_probe.hpp"
#include "photonflow/blocks/block_registry.hpp"

namespace photonflow {

OSAProbe::OSAProbe(const std::string &id, const json &params,
                   const json &nonideal)
    : BaseBlock(id, params, nonideal) {}

std::unordered_map<std::string, Signal>
OSAProbe::process(const std::unordered_map<std::string, Signal> &inputs,
                  SimContext & /*ctx*/) {
  // Get input signal for measurement capture
  auto it = inputs.find("opt_in");
  if (it == inputs.end()) {
    return {};
  }

  // Probes don't output anything - they are measurement points
  // The signal is captured via the graph execution logic
  return {};
}

std::optional<std::string> OSAProbe::port_type(const std::string &port) const {
  if (port == "opt_in")
    return "optical";
  return std::nullopt;
}

json OSAProbe::describe() const {
  return {{"ports", {{"opt_in", "optical"}}},
          {"spec",
           {{"params",
             {{"window",
               {{"type", "enum"},
                {"default", "hann"},
                {"options", json::array({"hann", "hamming", "blackman", "rect",
                                         "kaiser"})}}},
              {"ref", {{"type", "float"}, {"default", 1.0}, {"unit", ""}}},
              {"include_power", {{"type", "bool"}, {"default", false}}}}},
            {"nonideal", json::object()}}}};
}

REGISTER_BLOCK(OSAProbe, "OSAProbe");

} // namespace photonflow

File: backend/src/photonflow/blocks/measurement/osa_probe.py
"""Optical spectrum analyzer probe."""

from __future__ import annotations

from typing import Dict

from photonflow.blocks.base import BaseBlock, register_block
from photonflow.core.signal import Signal
from photonflow.core.sim import SimContext


@register_block("OSAProbe")
class OSAProbe(BaseBlock):
    """Optical Spectrum Analyzer (OSA) probe. Measures the optical power spectrum of a signal, displaying power vs. wavelength/frequency."""

    PORTS = {"opt_in": "optical"}
    SPEC = {
        "params": {
            "window": {
                "type": "enum",
                "default": "hann",
                "options": ["hann", "hamming", "blackman", "rect", "kaiser"],
            },
            "ref": {"type": "float", "default": 1.0, "unit": ""},
            "include_power": {"type": "bool", "default": False},
        },
        "nonideal": {},
    }

    def process(self, inputs: Dict[str, Signal], ctx: SimContext) -> Dict[str, Signal]:
        _ = inputs.get("opt_in")
        return {}
